internal/util/fencing.go

```diff
@@ -0,0 +1,163 @@
+)
+
+type NetworkFence struct {
+	Cidr      []string
+	Monitors  string
+	CreatedAt *timestamp.Timestamp
+}
+
+// addCephBlocklist adds an IP to ceph osd blocklist.
+func (nwFence *NetworkFence) addCephBlocklist(ctx context.Context, ip string, cr *Credentials) error {
+	arg := []string{
+		"--id", cr.ID,
+		"--keyfile=" + cr.KeyFile,
+		"-m", nwFence.Monitors,
+	}
+	// TODO: add blocklist till infinity.
+	// Currently, ceph does not provide the functionality to blocklist IPs
+	// for infinite time. As a workaround, add a blocklist for 5 YEARS to
+	// represent infinity from ceph-csi side.
+	// At any point in this time, the IPs can be unblocked by an UnfenceClusterReq.
+	// This needs to be updated once ceph provides functionality for the same.
+	blocklistTime := "157784760"
+	cmd := []string{"osd", "blocklist", "add", ip, blocklistTime}
+	cmd = append(cmd, arg...)
+	_, _, err := ExecCommand(ctx, "ceph", cmd...)
+	if err != nil {
+		return fmt.Errorf("failed to blocklist IP %q: %w", ip, err)
+	}
+	log.DebugLog(ctx, "blocklisted ip %s", ip)
+
+	return nil
+}
+
+// AddNetworkFence blocks access for all the IPs in the IP range mentioned via the CIDR block
+// using a network fence.
+func (nwFence *NetworkFence) AddNetworkFence(ctx context.Context, cr *Credentials) error {
+	// for each cidr block, convert it into a range of IPs so as to perform blocklisting operation.
+	for _, cidr := range nwFence.Cidr {
+		// fetch the list of IPs from a CIDR block
+		hosts, err := getIPRange(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to convert cidr block %s to corresponding IP range", cidr)
+		}
+
+		// add ceph blocklist for each IP in the range mentioned by the CIDR
+		for _, host := range hosts {
+			err = nwFence.addCephBlocklist(ctx, host, cr)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// getIPRange returns a list of IPs from the IP range
+// corresponding to a CIDR block.
+func getIPRange(cidr string) ([]string, error) {
+	var hosts []string
+	netIP, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, err
+	}
+	for ip := netIP.Mask(ipnet.Mask); ipnet.Contains(ip); inc(ip) {
+		hosts = append(hosts, ip.String())
+	}
+
+	return hosts, nil
+}
+
+// inc is an helper function for getIPRange for incrementing
+// IP values to return all IPs in a range.
+func inc(ip net.IP) {
+	for j := len(ip) - 1; j >= 0; j-- {
+		ip[j]++
+		if ip[j] > 0 {
+			break
+		}
+	}
+}
+
+type Cidrs []*fence.CIDR
+
+// CidrList converts a CIDR struct list to a list.
+func CidrList(cidrs Cidrs) ([]string, error) {
+	var cidrList []string
+	for _, cidr := range cidrs {
+		cidrList = append(cidrList, cidr.Cidr)
+	}
+	if len(cidrList) < 1 {
+		return nil, errors.New("the cidr cannot be empty")
+	}
+
+	return cidrList, nil
+}
+
+// removeCephBlocklist removes an IP from ceph osd blocklist.
+func (nwFence *NetworkFence) removeCephBlocklist(ctx context.Context, ip string, cr *Credentials) error {
```

Comment:
keep the receiver with a short name in all places it can be called as `nf`

Reply:
Done :+1: